libs/community/langchain_community/chat_message_histories/sql.py

```diff
@@ -80,36 +92,85 @@ def get_sql_model_class(self) -> Any:
         session_id_field_name: str = "session_id",
         custom_message_converter: Optional[BaseMessageConverter] = None,
     ):
-        self.connection_string = connection_string
-        self.engine = create_engine(connection_string, echo=False)
+        if connection_string:
+            logger.warning("connection_string is deprecated, use connection instead")
+            connection = connection_string
+            self.connection_string = connection_string
+        if isinstance(connection, str):
+            self.async_mode = async_mode
+            if async_mode:
+                self.async_engine = create_async_engine(
+                    connection, **(engine_args or {})
+                )
+            else:
+                self.engine = create_engine(url=connection, **(engine_args or {}))
+        elif isinstance(connection, Engine):
+            self.async_mode = False
+            self.engine = connection
+        elif isinstance(connection, AsyncEngine):
+            self.async_mode = True
+            self.async_engine = connection
+        else:
+            raise ValueError(
+                "connection should be a connection string or an instance of "
+                "sqlalchemy.engine.Engine or sqlalchemy.ext.asyncio.engine.AsyncEngine"
+            )
+
+        # To be consistent with others SQL implementations, rename to session_maker
+        self.session_maker: Union[scoped_session, async_sessionmaker]
+        if self.async_mode:
+            self.session_maker = async_sessionmaker(bind=self.async_engine)
+        else:
+            self.session_maker = scoped_session(sessionmaker(bind=self.engine))
+
         self.session_id_field_name = session_id_field_name
         self.converter = custom_message_converter or DefaultMessageConverter(table_name)
         self.sql_model_class = self.converter.get_sql_model_class()
         if not hasattr(self.sql_model_class, session_id_field_name):
             raise ValueError("SQL model class must have session_id column")
-        self._create_table_if_not_exists()
+        self._table_created = False
+        if not self.async_mode:
+            self._create_table_if_not_exists()
 
         self.session_id = session_id
-        self.Session = sessionmaker(self.engine)
 
     def _create_table_if_not_exists(self) -> None:
         self.sql_model_class.metadata.create_all(self.engine)
+        self._table_created = True
+
+    async def _acreate_table_if_not_exists(self) -> None:
+        if not self._table_created:
+            async with self.async_engine.begin() as conn:
+                await conn.run_sync(self.sql_model_class.metadata.create_all)
+            self._table_created = True
 
     @property
     def messages(self) -> List[BaseMessage]:  # type: ignore
         """Retrieve all messages from db"""
-        with self.Session() as session:
+        assert not self.async_mode, "This method must be called without async_mode"
```

Comment:
Could you push this assert statement inside `session_maker`?

Reply:
No. Because the 
```
with self.session_maker() as session:
```
is different with
```
async with self.session_maker() as session:
```

Reply:
Could you add something like this, so the asserts are always applied?


```python
    @contextlib.contextmanager
    def _make_sync_session(self) -> Session:
        """Make an async session."""
        if self.async_mode:
            raise ValueError(
                "Attempting to use a sync method in when async mode is turned on. "
                "Please use the corresponding async method instead."
            )
        with self.session_maker() as session:
            yield cast(Session, session)

    @contextlib.asynccontextmanager
    async def _make_async_session(self) -> AsyncSession:
        """Make an async session."""
        if not self.async_mode:
            raise ValueError(
                "Attempting to use an async method in when sync mode is turned on. "
                "Please use the corresponding async method instead."
            )
        async with self.session_maker() as session:
            yield cast(AsyncSession, session)
```